### ISSUE TITLE
Identifiers beginning with digits

### ISSUE THREAD

**Issue by wiztigers** [TypeCobolTeam/TypeCobol#50]
**BLOCKS ISSUE #23**

Seen in real-life code: `PERFORM 0000-INIT-STD`, as the first line of a PROCEDURE DIVISION.
Are identifier names allowed to begin with digits ?
If yes, in which context(s) ? Always ?

**Comment by osmedile** [comment]
You can write identifiers and paragraphs that start with digits.

Page 9 of Cobol V5.1 reference:
> A COBOL word is a character-string that forms a user-defined word, a
> system-name, or a reserved word. The maximum size of a COBOL user-defined
> word is 30 bytes. The number of characters that can be specified depends on the
> code page indicated by the compile-time locale.
> Except for arithmetic operators and relation characters, each character of a COBOL
> word is selected from the following set:
> * Latin uppercase letters A through Z
> * Latin lowercase letters a through z
> * digits 0 through 9
> * - (hyphen)
> * _ (underscore)
>
> The hyphen cannot appear as the first or last character in such words. The
> underscore cannot appear as the first character in such words. Most user-defined
> words (all except section-names, paragraph-names, priority-numbers, and
> level-numbers) must contain at least one alphabetic character. Priority numbers and
> level numbers need not be unique; a given specification of a priority-number or
> level-number can be identical to any other priority-number or level-number.
> In COBOL words (but not in the content of alphanumeric, DBCS, and national
> literals), each lowercase single-byte alphabetic letter is considered to be equivalent
> to its corresponding single-byte uppercase alphabetic letter.*

**Comment by smedilol** [comment]
As stated by Cobol reference:
The mantissa of a floating point always contains a decimal point (char '.' by default or char ',' if DECIMAL-POINT IS COMMA is specified).
So there's no confusion between a floating-point and a variable name.


Picture of a floating-point use almot the same syntax, except that char 'V' can replace the decimal point. The following picture clause are equivalent:
```Cobol    
       01 floatingPoint  pic -9v9(9)E-99.
       01 floatingPoint2 pic -9.9(9)E-99.
```
I can then write the following paragraph:
```Cobol
       testFloatingPoint.
           compute floatingPoint = 2.3E-13 + -3.444E-1.
```

**Comment by laurentprudhon** [comment]
In the current version, the Scanner has been implemented with an important goal in mind : to help the developers correct their spelling mistakes by giving them meaningful error messages.
In several situations, the Scanner recognizes an "almost correct" input : it can then continue analyzing the program, and output a nice error message : "you must correct this detail, but I think I have understood what you where trying to say".
For example, if a user enters the string 23E-2, the Scanner will recognize it as an attempt to input a floating point literal, but will print an error message saying that a period is mandatory in the mantissa and that the exponent must be on two chars, helping the user to correct his mistake 23.0E-02.
If we don't do this, 23E-2 will be recognized a user defined word, a strange error will occur in the Parser while trying to match an arithmetic expression, and the developer may have a hard time understandind what happened.
So I chose to preserve this behavior, and then I can not support user defined words of the form 23E--02.
We have a tradeoff to make here : to make a good decision, we must try to evaluate which one of the two situations will be more frequent / important for our developers.
